### PR TITLE
Fix issue 22306 - scope array variable should be stack allocated

### PR DESCRIPTION
**Pull request by dkorpel**
This includes a backend change @kinke @ibuclaw. I tried to make it a rewrite like in https://github.com/dlang/dmd/pull/11039, but didn't find a way to declare a temporary static array in the parent scope so it wouldn't immediately be destructed after the array literal assignment. So currently, I'm mirroring the way it's done with `scope Object o = new Object()` and keep an `onstack` variable.

### CODE REVIEW

**Comment by Geod24** [comment]
Hum, the real MVP here would be to support `scope int[] array = new int[](someRuntimeValue);`

**Comment by Geod24** [comment]
> You need to decide on a good size cap to prevent stack overflow, and dynamically switch at runtime.

static arrays already have a limit, how about re-using that ?

> It can't statically be `@nogc` either.

That's the whole point. Instead of dynamically switching at runtime, you just give a hard error. It's easy enough to check beforehand.

> I know two kinds of MVP

Actually I was thinking "Most Valuable Player" but not sure my usage was right :upside_down_face:

**Comment by dkorpel** [comment]
> static arrays already have a limit, how about re-using that ?

That limit is far above a typical stack size. I'm basically asking for stack overflow if I write this:

```D
void main() {
    ubyte[32*1024*1024] x; // Initializes 32 Mb of stack memory
}
```

But here, I don't want the compiler to irresponsibly move this GC allocation to the stack:
```D
void main() {
    scope ubyte[] x = new ubyte[](32*1024*1024);
}
```

> That's the whole point. Instead of dynamically switching at runtime, you just give a hard error. It's easy enough to check beforehand.

When I use the `new` operator, I assume free space on the heap is available to me, I don't want an unnecessary Out of Memory error.

> Actually I was thinking "Most Valuable Player" but not sure my usage was right :upside_down_face:

I think it's right, I'm just not used to hearing it in a software context where "Minimum Viable Product" is more prevalent.

File: changelog/scope-array-on-stack.dd [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+This can now be written in a single line:
+
+---
+void main() @nogc
+{
+    scope int[] arr = [10, 20, 30];
+}
+---
+
+This only applies to array elements that are Plain Old Data though, because proper destruction cannot be ensured when the array is not managed by the GC.
```
Comment (maxhaton) [inline changelog/scope-array-on-stack.dd:23]:
```suggestion
This only applies to array elements that are Plain Old Data, proper destruction cannot be ensured when the array is not managed by the GC.
```

**Comment by maxhaton** [comment]
A general framework for moving gc allocations either onto the stack or malloc and free or whatever would be good but requires a level of trust in the rest of the frontend that I don't really have

**Comment by maxhaton** [comment]
A test involving a scope array and a closure would be good, i.e. to make sure it ends up on the correct "stack"

**Comment by thewilsonator** [comment]
If you want GC promotions to the stack, use LDC

**Comment by nordlow** [comment]
Scoped allocation in cases such as

```d
void f()
{
    class C { int x; }
    scope c = new C(); // c is freed at the end of f()
}
```

already uses malloc+free under the hood. Can we please have arrays utilize the same optimization. Starting with POD-types element types only.

**Comment by nordlow** [comment]
> Hum, the real MVP here would be to support `scope int[] array = new int[](someRuntimeValue);`

Exactly my thought aswell.

Cases such as

```d
void f()
{
    class C { int x; }
    scope c = new C(); // c is freed at the end of f()
}
```

already uses malloc+free under the hood. Can we please have arrays utilize the same optimization? Starting with POD-types element types only, I guess.

**Comment by Geod24** [comment]
> already uses malloc+free under the hood.

No it doesn't. It's stack allocated, not `malloc`ed.

> If you want GC promotions to the stack, use LDC

It's an optimization, which is great, but has the major downsides of all optimizations: It cannot be enforced (easily & statically).

> That limit is far above a typical stack size.

What's a typical stack size ? There are a myriad of ways to control stack size (ulimit, fibers) and thus there's no "Right default".

> When I use the `new` operator, I assume free space on the heap is available to me, I don't want an unnecessary Out of Memory error.

Out of memory happens when an allocation can't be performed. Where the allocation is, that's an implementation detail. Remember we have different types of GC.

Also you can do this currently:
```D
class Foobar
{
    int[32*1024*1024 - 1] loads;
}

void main () @nogc
{
    scope f = new Foobar;
}
```

And yes it SEGV.

**Comment by nordlow** [comment]
> > already uses malloc+free under the hood.
> 
> No it doesn't. It's stack allocated, not `malloc`ed, and has been since D1.

Ok, sorry. My mistake. I'll overstrike it.

> ```d
> class Foobar
> {
>     int[32*1024*1024 - 1] loads;
> }
> 
> void main () @nogc
> {
>     scope f = new Foobar;
> }
> ```
> 
> And yes it SEGV.

Would be worthwhile looking into if GCC/Clang diagnoses that as an error statically.

**Comment by Geod24** [comment]
> whereas disallows

It doesn't disallow it. It just overflows the integer constant, leading to a negative size, and `clang` fails to correctly diagnose this, while GCC tells you exactly what the problem is. If you use `2*1024*1024*1024UL` it'll happily compile.

**Review by WalterBright — APPROVED** [review]

**Comment by WalterBright** [comment]
> Hum, the real MVP here would be to support scope int[] array = new int[](someRuntimeValue);

That is an important topic for discussion, but should not delay this PR.

**Comment by nordlow** [comment]
> It doesn't disallow it. It just overflows the integer constant, leading to a negative size, and `clang` fails to correctly diagnose this, while GCC tells you exactly what the problem is. If you use `2*1024*1024*1024UL` it'll happily compile.

Oops. Thanks.

**Comment by dkorpel** [comment]
> @dkorpel could you please follow this up with a spec PR?

https://github.com/dlang/dlang.org/pull/3442

**Pull request merged**